File: valid-anagram/flynn.cpp
/**
 * For length of given strings N,
 * 
 * Time complexity: O(N)
 * - iteration for given strings
 * 
 * Space complexity: O(1)
 * - the size of the container `count` is constant
 */

class Solution {
public:
    bool isAnagram(string s, string t) {
        int count[26] = {0};

        for (auto c : s) count[c - 'a']++;
        for (auto c : t) count[c - 'a']--;

        for (int i = 0; i < 26; i++) if (count[i]) return false;
        return true;
    }
};

File: valid-anagram/flynn.py
'''
For N, length of the given strings,

Time Complexity: O(N)
- first iteration: O(N)
- second iteration: O(N)

Space Compelxity: O(N)
- dictionaries for each strings: O(N)
'''

class Solution:
    def isAnagram(self, s: str, t: str) -> bool:
        if len(s) != len(t):
            return False

        count_s, count_t = {}, {}
        for i in range(len(s)):
            count_s[s[i]] = count_s.get(s[i], 0) + 1
            count_t[t[i]] = count_t.get(t[i], 0) + 1
        
        for c in count_s:
            if c not in count_t or count_s[c] != count_t[c]:
                return False
        return True
